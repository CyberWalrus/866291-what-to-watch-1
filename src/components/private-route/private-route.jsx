import React from "react";
import {Redirect, Route} from "react-router-dom";
import PropTypes from "prop-types";
import {getAuthorizationStatus} from "../../store/user/selectors.js";
import {connect} from "react-redux";
import RoutePath from "../../routes.js";

const PrivateRoute = ({
  component: Component,
  isAuthorizationRequired,
  rest,
  redirectPath = RoutePath.LOGIN,
  isAuthor = true,
  path
}) => {
  return (
    <Route
      {...rest}
      path={path}
      render={(props) =>
        isAuthorizationRequired === isAuthor ? (
          <Component {...props} />
        ) : (
          <Redirect
            to={{
              pathname: redirectPath,
              state: {from: props.location}
            }}
          />
        )
      }
    />
  );
};
PrivateRoute.propTypes = {
  rest: PropTypes.any,
  path: PropTypes.string,
  component: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.func
  ]).isRequired,
  isAuthorizationRequired: PropTypes.bool.isRequired,
  location: PropTypes.object.isRequired,
  isAuthor: PropTypes.bool,
  redirectPath: PropTypes.string
};

const mapStateToProps = (state, ownProps) =>
  Object.assign({}, ownProps, {
    isAuthorizationRequired: getAuthorizationStatus(state)
  });

export {PrivateRoute};

export default connect(mapStateToProps)(PrivateRoute);
