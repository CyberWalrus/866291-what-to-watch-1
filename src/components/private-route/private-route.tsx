import * as React from "react";
import {Redirect, Route} from "react-router-dom";
import {getAuthorizationStatus} from "../../store/user/selectors";
import {connect} from "react-redux";
import RoutePath from "../../routes";
import {StateApp} from "../../type/reducer";

interface PropsInsert {
  rest?: any,
  path?: string,
  component: any,
  location: object,
  isAuthor?: boolean,
  redirectPath?: string
}
interface PropsState {
  isAuthorizationRequired: boolean,
}
type Props = PropsInsert & PropsState;

const PrivateRoute = ({
  component: Component,
  isAuthorizationRequired,
  rest,
  redirectPath = RoutePath.LOGIN,
  isAuthor = true,
  path
}: Props) => {
  return (
    <Route
      {...rest}
      path={path}
      render={(props) =>
        isAuthorizationRequired === isAuthor ? (
          <Component {...props} />
        ) : (
          <Redirect
            to={{
              pathname: redirectPath,
              state: {from: props.location}
            }}
          />
        )
      }
    />
  );
};

const mapStateToProps = (state: StateApp, ownProps: Props): Props =>
  Object.assign({}, ownProps, {
    isAuthorizationRequired: getAuthorizationStatus(state)
  });

export {PrivateRoute};

export default connect(mapStateToProps)(PrivateRoute);
