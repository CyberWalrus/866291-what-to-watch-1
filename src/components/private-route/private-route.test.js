import React from "react";
import renderer from "react-test-renderer";
import {PrivateRoute} from "./private-route.jsx";
import {BrowserRouter, Switch} from "react-router-dom";

it(`PrivateRoute correctly renders after relaunch`, () => {
  const TestComponent = () => {
    return <div />;
  };
  const tree = renderer
    .create(
        <BrowserRouter>
          <Switch>
            <PrivateRoute
              component={TestComponent}
              isAuthorizationRequired={true}
              location={{}}
            />
          </Switch>
        </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
