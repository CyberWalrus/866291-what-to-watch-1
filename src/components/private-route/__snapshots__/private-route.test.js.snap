// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute correctly renders after relaunch 1`] = `<div />`;
